ergo-protocol-framework/src/box_spec.rs

```diff
@@ -0,0 +1,233 @@
+pub use ergo_lib::ast::Constant;
+use ergo_lib::chain::address::{Address, AddressEncoder, NetworkPrefix};
+pub use ergo_lib::chain::ergo_box::ErgoBox;
+pub use ergo_lib::chain::token::{TokenAmount, TokenId};
+use ergo_lib::serialization::serializable::SigmaSerializable;
+use ergo_lib::ErgoTree;
+use ergo_lib_wasm::box_coll::ErgoBoxes;
+use ergo_lib_wasm::ergo_box::ErgoBox as WErgoBox;
+use ergo_offchain_utilities::encoding::address_string_to_ergo_tree;
+use ergo_offchain_utilities::{ErgoAddressString, NanoErg};
+use serde_json::from_str;
+use std::ops::Range;
+use thiserror::Error;
+use wasm_bindgen::prelude::*;
+
+pub type Result<T> = std::result::Result<T, ProtocolFrameworkError>;
+
+#[derive(Error, Debug)]
+pub enum ProtocolFrameworkError {
+    #[error("The address of the box does not match the address.")]
+    InvalidAddress,
+    #[error("The number of Ergs held within the box is outside of the valid range.")]
+    InvalidErgsValue,
+    #[error("One of the token predicates failed for the provided box.")]
+    FailedTokenSpec,
+    #[error("One of the register predicates failed for the provided box.")]
+    FailedRegisterSpec,
+    #[error("The provided TokenId is invalid.")]
+    InvalidTokenId,
+    #[error("{0}")]
+    Other(String),
+}
+
+#[wasm_bindgen]
+#[derive(Clone)]
+pub struct TokenSpec {
+    value_range: Range<u64>,
```

Comment:
Since a valid range for token amounts is `1..i64::MAX` I suggest using `TokenAmount` newtype.

Reply:
IIRC a `TokenAmount` specifies an exact number though, right? The idea of the range is that a valid Box matching the spec can have any value within said range and still be considered valid.

Or are you recommending using `Range<TokenAmount>`?

Reply:
Yes, I meant `Range<TokenAmount>`.